### ISSUE TITLE
Upperbound version restrictions for Flutter SDK are ignored  

### ISSUE THREAD

**Issue by ivard** [flutter/flutter#90773]
## Steps to Reproduce

1. Specify a build environment for `flutter` with an upperbound version restriction in the project's `pubspec.yaml`.
```
environment:
  sdk: ">=2.13.0 <3.0.0"
  flutter: ">=2.2.0 <2.3.0"
```
2. Run `flutter pub get` to reload the `pubspec.yaml` using Flutter 2.2.2.
3. In `pubspec.lock` the upperbound version is now recorded differently:
```
sdks:
  dart: ">=2.13.0 <3.0.0"
  flutter: ">=2.2.0"
```
4. Now, when upgrading Flutter to version 2.5.1 (stable) and running `flutter pub get` again, no error message is visible that the build environment does not match.

## Logs

<details>
<summary>Logs</summary>

```
$ flutter pub get
Downloading Dart SDK from Flutter engine 91c9fc8fe011352879e3bb6660966eafc0847233...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  199M  100  199M    0     0  20.9M      0  0:00:09  0:00:09 --:--:-- 28.8M
Building flutter tool...
  ╔════════════════════════════════════════════════════════════════════════════╗
  ║ A new version of Flutter is available!                                     ║
  ║                                                                            ║
  ║ To update to the latest version, run "flutter upgrade".                    ║
  ╚════════════════════════════════════════════════════════════════════════════╝


Downloading package sky_engine...                                  355ms
Downloading flutter_patched_sdk tools...                           249ms
Downloading flutter_patched_sdk_product tools...                   223ms
Downloading linux-x64 tools...                                   1,183ms
Downloading linux-x64/font-subset tools...                         149ms
Running "flutter pub get" in project...                       1,121ms
$ flutter upgrade
Upgrading Flutter to 2.5.1 from 2.2.2 in /home/ivar/Android/flutter...
Downloading Dart SDK from Flutter engine b3af521a050e6ef076778bcaf16e27b2521df8f8...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  2  221M    2 5518k    0     0  10.1M      0  0:00:21 --:--:--  0:00:21 10.1M
 14  221M   14 31.4M    0     0  20.5M      0  0:00:10  0:00:01  0:00:09 20.5M
 26  221M   26 58.3M    0     0  23.0M      0  0:00:09  0:00:02  0:00:07 23.0M
 38  221M   38 84.6M    0     0  23.9M      0  0:00:09  0:00:03  0:00:06 23.9M
 50  221M   50  111M    0     0  24.5M      0  0:00:09  0:00:04  0:00:05 24.5M
 60  221M   60  133M    0     0  24.1M      0  0:00:09  0:00:05  0:00:04 25.6M
 72  221M   72  159M    0     0  24.4M      0  0:00:09  0:00:06  0:00:03 25.6M
 83  221M   83  184M    0     0  24.4M      0  0:00:09  0:00:07  0:00:02 25.2M
 93  221M   93  205M    0     0  23.8M      0  0:00:09  0:00:08  0:00:01 23.7M
100  221M  100  221M    0     0  24.0M      0  0:00:09  0:00:09 --:--:-- 23.5M
Building flutter tool...

Upgrading engine...
Downloading package sky_engine...                                  376ms
Downloading flutter_patched_sdk tools...                           297ms
Downloading flutter_patched_sdk_product tools...                   248ms
Downloading linux-x64 tools...                                   1,274ms
Downloading linux-x64/font-subset tools...                         351ms

Flutter 2.5.1 • channel stable • https://github.com/flutter/flutter.git
Framework • revision ffb2ecea52 (9 days ago) • 2021-09-17 15:26:33 -0400
Engine • revision b3af521a05
Tools • Dart 2.14.2

Running flutter doctor...
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.5.1, on Ubuntu 20.04.3 LTS 5.4.0-81-generic, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2021.2)
[✓] IntelliJ IDEA Ultimate Edition (version 2021.1)
[✓] Connected device (1 available)

• No issues found!
$ flutter pub get
Running "flutter pub get" in project...                       1,808ms
The plugins `flutter_jailbreak_detection, flutter_privacy_screen` use a deprecated version of the Android embedding.
To avoid unexpected runtime failures, or future build failures, try to see if these plugins support the Android V2 embedding. Otherwise, consider removing them since a
future release of Flutter will remove these deprecated APIs.
If you are plugin author, take a look at the docs for migrating the plugin to the V2 embedding: https://flutter.dev/go/android-plugin-migration.
```

```
[✓] Flutter (Channel stable, 2.2.2, on Linux, locale en_US.UTF-8)
    • Flutter version 2.2.2 at /home/ivar/Android/flutter
    • Framework revision d79295af24 (4 months ago), 2021-06-11 08:56:01 -0700
    • Engine revision 91c9fc8fe0
    • Dart version 2.13.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /home/ivar/Android/Sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /home/ivar/Android/Sdk
    • Java binary at: /snap/android-studio/101/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /snap/android-studio/101/android-studio
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2)
    • IntelliJ at /snap/intellij-idea-ultimate/311
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.4746.57

[✓] IntelliJ IDEA Ultimate Edition (version 2021.1)
    • IntelliJ at /snap/intellij-idea-ultimate/289
    • Flutter plugin version 55.0.5
    • Dart plugin version 211.6693.108

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 94.0.4606.61

• No issues found!
```

</details>

**Comment by stuartmorgan** [comment]
@mit-mit @csells Do we have documentation about this behavior anywhere?

**Comment by ivard** [comment]
The documentation of the SDK version constraints can be found [here](https://dart.dev/tools/pub/pubspec#sdk-constraints). 

I also tried to specify a specific SDK version, like use exactly version 2.2.2, but that one is also converted to `>= 2.2.2`.

**Comment by TahaTesser** [comment]
Hi @ivard 
I can reproduce the issue in Fluttter. However, this also reproduces in a dart project with Flutter SDK and `dart pub get` 


`dart create hello`
```yaml
name: hello
description: A simple command-line application.
version: 1.0.0
# homepage: https://www.example.com

environment:
  sdk: '>=2.12.0 <3.0.0'
  flutter: ">=2.2.0 <2.3.0"


# dependencies:
#   path: ^1.8.0

dev_dependencies:
  lints: ^1.0.0
```

```lock
# Generated by pub
# See https://dart.dev/tools/pub/glossary#lockfile
packages:
  lints:
    dependency: "direct dev"
    description:
      name: lints
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.1"
sdks:
  dart: ">=2.12.0 <3.0.0"
  flutter: ">=2.2.0"
```

This looks like an intended behavior, these [changes](https://github.com/dart-lang/pub/pull/2780/files#diff-157abc64e6b3253bb39fc124470db1f94ec478472d7223852864b1a41058fbb0R348) are made in https://github.com/dart-lang/pub/pull/2780 and https://github.com/dart-lang/pub/issues/2696
Change

**Comment by stuartmorgan** [comment]
That's what my question above was regarding. It should certainly be documented, it's just a question of where, and if it already is somewhere.

**Comment by ivard** [comment]
Was there a special reason why it was deprecated? We see a lot of contributers struggling with finding the right Flutter version to build our project. Having a newer SDK version of Flutter might also lead to issues, for example if a dependency has no support for the newer Flutter version yet. 

Of course we can also solve this by mentioning it in the project's README, so it is not strictly necessary. However, enforcing a Flutter SDK version in the `pubspec.yaml` appeared to me as a nice extra way to help people setting up a project.

**Comment by christopherfujino** [comment]
This looks like the tracking issue with context: https://github.com/flutter/flutter/issues/68143

**Comment by stuartmorgan** [comment]
> Having a newer SDK version of Flutter might also lead to issues, for example if a dependency has no support for the newer Flutter version yet.

Flutter, unlike the rest of the ecosystem, does not follow semver, so there is no way of setting a meaningful upper constraint in advance. Is the use case here an actively maintained project that doesn't support *current* versions of Flutter?

**Comment by christopherfujino** [comment]
> > Having a newer SDK version of Flutter might also lead to issues, for example if a dependency has no support for the newer Flutter version yet.
> 
> Flutter, unlike the rest of the ecosystem, does not follow semver, so there is no way of setting a meaningful upper constraint in advance. Is the use case here an actively maintained project that doesn't support _current_ versions of Flutter?

It would be nice to be able to pick up patch versions of the current stable, something like:

```
sdks:
  flutter: ">=2.5.0 <2.6.0"
```

**Comment by stuartmorgan** [comment]
> It would be nice to be able to pick up patch versions of the current stable

In what scenario though? What's the real-world use case? Is the goal to prevent people from using anything but the current stable version preemptively on the off-chance that a newer version might not work? Or is it to prevent people from using versions from other channels that already have breaking changes (rather than fixing them)?

**Comment by stuartmorgan** [comment]
Also, is the use case here specifically apps? Treating them differently than libraries in this context might make sense.

**Comment by ivard** [comment]
Yes right. In our case it concerns a mobile app. We get issues like [this one](https://github.com/privacybydesign/irmamobile/issues/19), because people are using the wrong SDK version.

This issue was also related to the fact that we provided an outdated `Makefile`, but even then a clear error that someone is using a wrong SDK version would help.

**Comment by fertrig** [comment]
We are also running into this issue. Flutter `3.8.0-0.0.pre` introduced a breaking change for us.

Before Flutter `3.8.0-0.0.pre` we could call `WidgetsBinding.attachRootWidget` like this:

```dart
binding.attachRootWidget(MyApp());
```

However, as of Flutter `3.8.0-0.0.pre`, we have to call it like this:
```dart
binding.attachRootWidget(binding.wrapWithDefaultView(MyApp()));
```

I reported this issue. See #121180.

Using an upper bound constraint on the flutter sdk version would have been useful. I wanted to ship a version of our package using an upper bound on the flutter sdk, like this:

```yaml
name: monarch
version: 3.0.1
...

environment:
  sdk: '>=2.12.0 <4.0.0'
  flutter: '>=2.12.0-4.1.pre <3.8.0-0.0.pre'
```

And then another version of our package using `3.8.0-0.0.pre` as the lower bound:
```yaml
name: monarch
version: 3.3.0
...

environment:
  sdk: '>=2.12.0 <4.0.0'
  flutter: '>=3.8.0-0.0.pre'
```

I expected the upper bound constraint to work. However, it is ignored. Users that upgrade to Flutter 3.8 may end up using an incompatible version of our package.

**Comment by stuartmorgan** [comment]
> I wanted to ship a version of our package using an upper bound on the flutter sdk, like this: [...]
> And then another version of our package using `3.8.0-0.0.pre` as the lower bound:

The upper bound wouldn't really do anything useful in that scenario; package resolution will give the highest viable version, so clients on a version compatible with your 3.3.0 would get that, and everyone else would get 3.0.1, even without the upper bound.

> Users that upgrade to Flutter 3.8 may end up using an incompatible version of our package.

That's true regardless. Presumably you had at some point in the past released a version 3.0.0 without an upper bound, which someone updating to Flutter 3.8.0 can continue to use regardless of what else you release. To prevent that you would need [retroactive constraints](https://github.com/dart-lang/pub-dev/issues/5587).

Which is exactly why the current behavior exists: you had no way to know when you published 3.0.0 what upper bound it would need, because Flutter doesn't use semver, which makes upper bounds useless in packages.

**Comment by ivard** [comment]
That's true, but if people download the Flutter SDK, they always get the latest stable version. This means that if a project is not compatible with that version yet, they will get into trouble. There is no nice way to communicate this now. The only thing we can do is to mention the version issue in a README and then hope that people see it.

**Comment by stuartmorgan** [comment]
@ivard I was responding to a comment about a package; your use case is an app, which as I said above is very different.

**Comment by fertrig** [comment]
@stuartmorgan:

> The upper bound wouldn't really do anything useful in that scenario; package resolution will give the highest viable version, so clients on a version compatible with your 3.3.0 would get that, and everyone else would get 3.0.1, even without the upper bound.

I see what you are saying. That makes sense now. I don't need an upper bound in this scenario.

> ... because Flutter doesn't use semver...

What do you mean by that? It certainly looks like Flutter uses semver. Also, is it ok to use the package [pub_semver](https://pub.dev/packages/pub_semver) to compare Flutter versions?

**Comment by stuartmorgan** [comment]
> > ... because Flutter doesn't use semver...
> 
> What do you mean by that?

I mean that Flutter versioning doesn't follow the requirements of semver. Most importantly in this context, it doesn't follow this one:

>> Major version X (X.y.z | X > 0) MUST be incremented if any backwards incompatible changes are introduced to the public API.

which is why there's no way to ever set a useful upper bound for Flutter in advance, only after the fact. Unlike ecosystem packages that follow semver, where (in theory at least), setting a `^` range on packages avoids later incompatibility.

> It certainly looks like Flutter uses semver.

If Flutter followed semver, only Flutter 2 and Flutter 3 would have any entries [here](https://docs.flutter.dev/release/breaking-changes).

**Comment by jaysephjw** [comment]
+1 for the app use case, specifically setting an exact version.

- Have run into issues where a teammate pulls in newer version w breaking change. 
- Want to ensure builds use an exact sdk version
- sdk version upgrades are intentional. 

pubspec seems the best place to specify this, without introducing extra tooling.

**Comment by nav9** [comment]
Hope a solution comes through soon. I just encountered this issue, where even after specifying `sdk: '>=2.12.0 <4.0.0'` for using `leveldb: ^7.0.0`, I get the error `The current Dart SDK version is 3.4.1.
Because timers_cascade depends on leveldb >=2.0.1 which doesn't support null safety, version solving failed.
The lower bound of "sdk: '<2.0.0 or >=2.0.0 <3.0.0'" must be 2.12.0 or higher to enable null safety.`.
  
Looking forward to a day when Users would no longer have to specify such things, and an AI would just select and download the necessary dependencies.

**Comment by sigurdm** [comment]
This is fixed by https://github.com/dart-lang/pub/pull/4595 (at least for root-packages) from dart 3.9 onward

Same as https://github.com/flutter/flutter/issues/95472
